### DISCUSSION TITLE
Multiple 'allowedVersions' settings for the same package pattern

### DISCUSSION THREAD

**Discussion by ablekhman** [renovatebot/renovate#9346] [Request Help]
Is it possible to have multiple `packageRules` entries that match the same package pattern, but specify different `allowedVersions` settings?

For example:
```json
"packageRules": [
    {
      "matchPackagePatterns": [ "^some.group" ],
      "allowedVersions": "<regex here>"
    },
    {
      "matchPackagePatterns": [ "^some.group" ],
      "allowedVersions": "< 1.2.0"
    }
]
```

**Comment by viceice** [comment]
Sure, but last rule wins. Please describe what you like to achieve, so we can help to configure it with right value

**Reply by ablekhman** [reply]
I would like to combine all of them, so all `allowedVersions` should satisfy before update is accepted. For example, I would like to have only numeric versions (`"allowedVersions": "/^[0-9]+\\.[0-9]+\\.[0-9]+(\\.[0-9]+)?$/"`) AND also versions lower than 6.x (`"allowedVersions": "< 6.0"`).

I am not sure how to enable both conditions for the same package pattern.

**Comment by viceice** [comment]
`"allowedVersions": "/^[0-5]\\.[0-9]+\\.[0-9]+(\\.[0-9]+)?$/"` ?

**Reply by ablekhman** [reply]
Well, in this particular example it will work, but I am after a generic solution. The settings I try to combine are:

* `"allowedVersions": "!/-[0-9a-f]{6,10}$/"` - disable certain versions with hash values at the end
* `"allowedVersions": "< 999.0.0"` - only allow versions bellow 999.0 (higher versions have special meaning and should not be used in certain projects)

**Reply by viceice** [reply]
`"allowedVersions": "/^[1-8]?[0-9]{1,2}\\.[0-9]+\\.[0-9]+\\.[0-9]+(-(dev|beta)(-\\d+)?)?$/"` something like this?

Please provide some more good and bad sample versions to improve the regex to your needs

**Reply by ablekhman** [reply]
My point is that I need *both*: a pattern and a range of versions. Solving both requirements with a regex is unwieldy. Even if particular regex works, it quickly becomes cumbersome and complex. And what if tomorrow I need more than two conditions? The regex will be unreadable and unmaintainable.

What if I specify one requirement for a broader group and the second for a narrower group? Like this:

```json
"packageRules": [
    {
      "matchPackagePatterns": [ "^some.group" ],
      "allowedVersions": "<regex here>"
    },
    {
      "matchPackagePatterns": [ "^some.group:package.name" ],
      "allowedVersions": "< 1.2.0"
    }
]
```

Can I expect that `some.group:package.name.*` dependencies will comply with both rules?
If not, then will `matchPackageNames` instead of `matchPackagePatterns` help with enabling multiple version requirements for the same package?

**Reply by viceice** [reply]
They won't get merged, last matching rule wins.

If you really need multiple `allowedVersions` feel free to open a new feature request. But be precise how it should work under different scenarios like multiple matching rules.

**Comment by jwnmulder** [comment]
Background: We have a lot of projects using maven for which Red Hat provides patched versions for lots of artifacts. As an example see this dependency: https://mvnrepository.com/artifact/commons-collections/commons-collections/3.2.1.redhat-7

Ideally we would like to configure Renovate in such a way that it has the following behavior

* artifacts that do not include `.redhat-` should not be updated to a version that includes `.redhat-`
* artifacts that do include a vendor classifier should keep it

The configuration we ended up with (simplified version)

```json
    {
      "managers": [
        "maven"
      ],
      "matchCurrentVersion": "!/[\\.-]redhat-\\d+$/",
      "allowedVersions": "!/[\\.-]redhat-\\d+$/"
    },
    {
      "managers": [
        "maven"
      ],
      "matchCurrentVersion": "/[\\.-]redhat-\\d+$/",
      "allowedVersions": "/[\\.-]redhat-\\d+$/"
    }
```

This is similar to the solution proposed in https://github.com/renovatebot/renovate/issues/5303#issuecomment-619448928 and has worked pretty well.

One problem we are having is that Renovate is creating PRs for commons-collections version '20040616'. This confused us as #7837 contained a fix for this. After reading the above it does make sense now. The question however, how can we best solve this?

Some ideas we have:

*Idea 1*: Create a custom version regex for each dependency to capture `.redhat-*` as a classifier
*  Lots of work and difficult to maintain as there are 100+ such dependencies

*Idea 2*: Keep the configuration above, but in addition create combined `allowedVersion` rules for each package found in [workarounds.ts](https://github.com/renovatebot/renovate/blob/main/lib/config/presets/internal/workarounds.ts)
* In this case there will be more than 2 packagerules required as redhat uses multiple vendor classifiers. Some examples: `redhat`, `jboss`, `jobssas`, etc.
* A downside is that we have to keep an eye on [workarounds.ts](https://github.com/renovatebot/renovate/blob/main/lib/config/presets/internal/workarounds.ts). When new entries are added we have to create new combined packagerules

Hopefully it gives some insight in what we would like to use `allowedVersions` for. Any pointers or ideas are welcome.

**Reply by ahoehma** [reply]
Is there a "final" solution for that scenario? I need to exclude ALL redhat versions and ALL "old" commons- versions.
